rfc-117-containerised-runtime.md

```diff
@@ -0,0 +1,68 @@
+
+GOV.UK has already containerised its development environment ([rfc-106](https://www.github.com/alphagov/govuk-rfcs/106)), a containerised runtime will narrow the gap between development and production.
+
+### Benefits of this approach
+
+The most immediate benefit of this approach is that it will allow us to incrementally upgrade the operating system version. By moving the applications into containers, GOV.UK can enable a much lower risk upgrade path.
+
+Moving to a container based hosting environment will allow us to remove much of our legacy Puppet and Fabric code. This code is hard to maintain, and hard to hire people with the experience required to improve and upgrade. With industry and government overwhelmingly moving towards containerised platforms, it should be much easier to hire people with these skills.
+
+Containerised hosting environments can generally provide much better value for money in terms of resource usage, both because many small applications can be “bin-packed” onto the same infrastructure, and because it is much easier to autoscale applications.
+
+Running containers in production allows for a reduction in the difference between development environments and production. This can lead to bugs being found more quickly, and faster development cycles.
+
+Other improvements to GOV.UK’s development lifecycle (such as moving to continuous deployment) will be easier to implement on a container based infrastructure than they would in the current virtual machine based infrastructure.
+
+### Risks of this approach
+
+GOV.UK has not yet estimated the length of time this work will take, our current assumption is that it will take much less than 2 years to complete. There is a risk that the “unknown unknowns” GOV.UK discovers will prevent GOV.UK from finishing the containerisation work before support ends for Trusty. If this happens, GOV.UK may need to refocus our efforts to upgrade from Trusty in the existing infrastructure.
+
+GOV.UK infrastructure has been in a transitory state for several years and containerisation will prolong this state. In addition in the short term GOV.UK will be adding complexity making supporting the platform hard for RE-GOV.UK and 2ndline.
+
+Focusing on containerisation will detract from work to improve other parts of the platform.
+
+GOV.UK has a significant amount of infrastructure outside of the applications themselves. In moving to a container based hosting platform, it’s possible that some existing behaviour may be missed, resulting in bugs.
+### Risks of not doing this
+
+GOV.UK reduces its options of how to mitigate Trusty reaching the end of support so the only option left is upgrading Ubuntu. This is potentially wasted work as once the upgrade is done GOV.UK will want to consider if running VMs is best for GOV.UK.
```

Comment:
Shouldn't this RFC be the consideration that "is running VMs is best for GOV.UK", rather than an upgrade path for Trusty?

Reply:
Could you expand on this point a bit?

I think the RFC does aim to answer "is running VMs best for GOV.UK?", broadly with "No, it would be best for GOV.UK to run containers on a managed platform of some kind".

A big part of the reason for that is the high difficulty of upgrading the OS on our existing VM estate though (in my opinion).

Reply:
Ah yes, sorry I should have wrote "should we containerise GOV.UK?" instead of "is running VMs is best for GOV.UK?".

but essentially I feel this RFC has been framed as solving issue of the expiring LTS of Trusty and us needing an upgrade path, rather than a more wholistic look at why we'd move containers. Maybe the summary paragraph (line 5) needs broadening? 

This line (58) suggests that we'd reconsider containerisation if this RFC was rejected. (Sorry there's a lot of double negative logic going on here)

Reply:
I've suggested an amendment to line 5 to make it a bit broader.


I think this line (58) is trying to say "A risk of not doing this [move to a containerised platform] is that we'll have to update Ubuntu the hard way (i.e. with puppet), which will be a big chunk of effort that brings limited benefit" - is that any clearer? Or perhaps it's less clear?

Reply:
Yup! Makes more sense!

Reply:
I've tried to clarify this a bit in 347e7a64a410ad5abb1d0cdea1c4efca419cf65b - I didn't quite use the wording above, but hopefully it's a bit clearer now.

Let me know if it still doesn't make sense :)